helix-view/src/document.rs

```diff
@@ -397,30 +398,111 @@ impl Document {
 
     /// The same as [`format`], but only returns formatting changes if auto-formatting
     /// is configured.
-    pub fn auto_format(&self) -> Option<impl Future<Output = LspFormatting> + 'static> {
-        if self.language_config()?.auto_format {
-            self.format()
+    pub fn auto_format(
+        &mut self,
+        view_id: ViewId,
+    ) -> anyhow::Result<Option<impl Future<Output = LspFormatting> + 'static>> {
+        let config = match self.language_config() {
+            Some(config) => config,
+            None => return Ok(None),
+        };
+
+        if config.auto_format {
+            self.format(view_id)
         } else {
-            None
+            Ok(None)
         }
     }
 
     /// If supported, returns the changes that should be applied to this document in order
     /// to format it nicely.
-    pub fn format(&self) -> Option<impl Future<Output = LspFormatting> + 'static> {
-        let language_server = self.language_server()?;
+    pub fn format(
+        &mut self,
+        view_id: ViewId,
+    ) -> anyhow::Result<Option<impl Future<Output = LspFormatting> + 'static>> {
+        if let Some(formatter) = self.language_config().and_then(|c| c.formatter.as_ref()) {
+            use std::process::{Command, Stdio};
+            match formatter.type_ {
+                FormatterType::Stdio => {
+                    use std::io::Write;
+                    let mut process = Command::new(&formatter.command)
+                        .args(&formatter.args)
+                        .stdin(Stdio::piped())
+                        .stdout(Stdio::piped())
+                        .stderr(Stdio::piped())
+                        .spawn()
+                        .map_err(|e| {
+                            anyhow!("Failed to format with {}. {}", formatter.command, e)
+                        })?;
+
+                    {
+                        let mut stdin = process.stdin.take().unwrap();
+                        stdin
+                            .write_all(&self.text().bytes().collect::<Vec<u8>>())
+                            .unwrap();
```

Comment:
Unwraps should be avoided where possible, and since we have Result as return type `?` can be used instead.